src/core/thread/network_data_local.cpp

```diff
@@ -111,29 +151,74 @@ Error Local::AddOnMeshPrefix(const OnMeshPrefixConfig &aConfig)
     }
 #endif
 
-    error =
-        AddPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeBorderRouter, aConfig.mPreference, flags, aConfig.mStable);
-
-exit:
-    return error;
+    return AddPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeBorderRouter, aConfig.mPreference, flags,
+                     aConfig.mStable);
 }
 
 Error Local::RemoveOnMeshPrefix(const Ip6::Prefix &aPrefix)
 {
     return RemovePrefix(aPrefix, NetworkDataTlv::kTypeBorderRouter);
 }
 
+Error Local::ValidateHasRoutePrefix(const ExternalRouteConfig &aConfig)
+{
+    Error error = kErrorNone;
+
+    // Validate Prefix
+    VerifyOrExit((aConfig.GetPrefix().GetLength() > 0) && aConfig.GetPrefix().IsValid(), error = kErrorInvalidArgs);
+
+    // Validate Preference
+    switch (aConfig.mPreference)
+    {
+    case OT_ROUTE_PREFERENCE_LOW:
+    case OT_ROUTE_PREFERENCE_MED:
+    case OT_ROUTE_PREFERENCE_HIGH:
+        break;
+    default:
+        ExitNow(error = kErrorInvalidArgs);
+    }
+
+    // Check if address already configured
+    VerifyOrExit(!aConfig.GetPrefix().ContainsPrefix(Get<Mle::MleRouter>().GetMeshLocalPrefix()),
+                 error = kErrorAlready);
+
+    // Verify network data contains enough space
+    VerifyOrExit(CheckBufferSizeForPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeHasRoute) == kErrorNone,
+                 error = kErrorInvalidArgs);
+
+exit:
+    return error;
+}
+
 Error Local::AddHasRoutePrefix(const ExternalRouteConfig &aConfig)
 {
-    return AddPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeHasRoute, aConfig.mPreference, /* aFlags */ 0,
-                     aConfig.mStable);
+    Error error = kErrorNone;
+
+    SuccessOrExit(error = ValidateHasRoutePrefix(aConfig));
+
+    SuccessOrExit(
+        error = AddPrefix(aConfig.GetPrefix(), NetworkDataTlv::kTypeHasRoute, aConfig.mPreference, 0, aConfig.mStable));
+
+exit:
+    return error;
 }
 
 Error Local::RemoveHasRoutePrefix(const Ip6::Prefix &aPrefix)
 {
     return RemovePrefix(aPrefix, NetworkDataTlv::kTypeHasRoute);
 }
 
+Error Local::CheckBufferSizeForPrefix(const Ip6::Prefix &aPrefix, NetworkDataTlv::Type aSubTlvType)
+{
+    uint8_t subTlvLength;
+
+    subTlvLength = (aSubTlvType == NetworkDataTlv::kTypeBorderRouter)
+                       ? sizeof(BorderRouterTlv) + sizeof(BorderRouterEntry)
+                       : sizeof(HasRouteTlv) + sizeof(HasRouteEntry);
+
+    return CanInsert(sizeof(PrefixTlv) + aPrefix.GetBytesSize() + subTlvLength) ? kErrorNone : kErrorNoBufs;
```

Comment:
`Local::AddPrefix` actually always remove the same prefix (if exist) before adding. So, we should consider that for calculating the space required?